codex/utils/options.nim

```diff
@@ -0,0 +1,61 @@
+import macros
+import strutils
+import pkg/questionable
+import pkg/questionable/operators
+
+export questionable
+
+proc `as`*[T](value: T, U: type): ?U =
+  ## Casts a value to another type, returns an Option.
+  ## When the cast succeeds, the option will contain the casted value.
+  ## When the cast fails, the option will have no value.
+  when value is U:
+    return some value
+  elif value is ref object:
+    if value of U:
+      return some U(value)
+
+Option.liftBinary `as`
+
+# Template that wraps type with `Option[]` only if it is already not `Option` type
+template WrapOption*(input: untyped): type =
+  when input is Option:
+    input
+  else:
+    Option[input]
+
+
+macro createType(t: typedesc): untyped =
+  var objectType = getType(t)
+
+  # Work around for https://github.com/nim-lang/Nim/issues/23112
+  while objectType.kind == nnkBracketExpr and objectType[0].eqIdent"typeDesc":
+    objectType = getType(objectType[1])
+
+  expectKind(objectType, NimNodeKind.nnkObjectTy)
+  var fields = nnkRecList.newTree()
+
+  # Generates the list of fields that are wrapped in `Option[T]`.
+  # Technically wrapped with `WrapOption` which is template used to prevent
+  # re-wrapping already filed which is `Option[T]`.
+  for field in objectType[2]:
+    let fieldType = getTypeInst(field)
+    let newFieldNode =
+          nnkIdentDefs.newTree(ident($field), nnkCall.newTree(ident("WrapOption"), fieldType), newEmptyNode())
+
+    fields.add(newFieldNode)
+
+  # Creates new object type T with the fields lists from steps above.
+  let tSym = genSym(nskType, "T")
+  nnkStmtList.newTree(
+      nnkTypeSection.newTree(
+        nnkTypeDef.newTree(tSym, newEmptyNode(), nnkObjectTy.newTree(newEmptyNode(), newEmptyNode(), fields))
+      ),
+      tSym
+  )
```

Comment:
```diff
+proc addSerializePragma(typeNode: NimNode): NimNode =
+  nnkPragmaExpr.newTree(
+    typeNode,
+    nnkPragma.newTree(
+      newIdentNode("serialize")
+    )
+  )

+ let tSymAnnotated = tSym.addSerializePragma()
  let tree = nnkStmtList.newTree(
      nnkTypeSection.newTree(
-        nnkTypeDef.newTree(tSym, newEmptyNode(), nnkObjectTy.newTree(newEmptyNode(), newEmptyNode(), fields))
+        nnkTypeDef.newTree(tSymAnnotated, newEmptyNode(), nnkObjectTy.newTree(newEmptyNode(), newEmptyNode(), fields))
      ),
      tSym
  )
```

Reply:
Nice! Did you test it? Does it work? Would you be willing to create PR for it?
IMHO I would just not be adding the pragma automatically to all the fields. Either, it should be smart enough to detect that the pragma was present on the original field and only then add it to the new type, or a simpler approach could be adding some flag to the `Optionalize()` which would enable this behavior.